closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -1021,10 +1046,11 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_DigestComputeAnd
 
     (*env)->ReleasePrimitiveArrayCritical(env, digest, digestNative, 0);
 
-    (*OSSL_MD_CTX_reset)(context->ctx);
-
-    if (1 != (*OSSL_DigestInit_ex)(context->ctx, context->digestAlg, NULL)) {
+    // Reset the message digest context to the original context. We are then ready to perform
+    // digest operations again using a copy of this cached context.
+    if (1 != (*OSSL_MD_CTX_copy_ex)(context->ctx, context->cachedInitalizedDigestContext)) {
         printErrors();
+        Java_jdk_crypto_jniprovider_NativeCrypto_DigestDestroyContext(env, thisObj, c);
```

Comment:
Won't this call (and the one on line 1079) potentially lead to releasing the context and the SSL objects two (or more) times? How would the Java code know *not* to destroy the context?

Reply:
I added code to nullify `context->ctx` and `context->cachedInitializedDigestContext` right after they are freed to avoid potential double frees for these values.

I think you are right that the `context` itself could potentially be double freed. I was able to prevent the double free of `context->ctx` and `context->cachedInitializedDigestContext` but not the `context` itself. For this reason I ended up changing these lines ( and the former line 1079 ) to just free and nullify `context->ctx` and `context->cachedInitializedDigestContext`, I no longer call the DigestDestroyContext for the exact reason you point out that a double free on the context itself could occur.

The Java code knows when to free these contexts by making use of a [Cleaner](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ref/Cleaner.html) which is registered right after a context is created successfully [at this location in the NativeDigest.java class](https://github.ibm.com/runtimes/openj9-openjdk-jdk/blob/c38252fc774e577f2b94a50612ba4193ef6ea75f/closed/src/java.base/share/classes/sun/security/provider/NativeDigest.java#L95).  From that point the long value ( which represents a pointer to the context ) is monitored by the Java code cleaner. Once that context becomes unreachable from a Java perspective then the  [DigestDestroyContext method](https://github.ibm.com/runtimes/openj9-openjdk-jdk/blob/c38252fc774e577f2b94a50612ba4193ef6ea75f/closed/src/java.base/share/classes/sun/security/provider/NativeDigest.java#L77) is called by the cleaners run method to cleanup the context by calling the JNI method `Java_jdk_crypto_jniprovider_NativeCrypto_DigestDestroyContext`.